Firestore/core/src/model/object_value.cc

```diff
@@ -0,0 +1,314 @@
+
+  for (const FieldPath& path : field_mask) {
+    if (!parent.IsImmediateParentOf(path)) {
+      // Insert the accumulated changes at this parent location
+      google_firestore_v1_MapValue* parent_map = ParentMap(parent);
+      ApplyChanges(parent_map, upserts, deletes);
+      upserts.clear();
+      deletes.clear();
+      parent = path.PopLast();
+    }
+
+    absl::optional<google_firestore_v1_Value> value = data.Get(path);
+    if (value) {
+      upserts.emplace(path.last_segment(), *value);
+    } else {
+      deletes.insert(path.last_segment());
+    }
+  }
+
+  google_firestore_v1_MapValue* parent_map = ParentMap(parent);
+  ApplyChanges(parent_map, upserts, deletes);
+}
+
+void MutableObjectValue::Delete(const FieldPath& path) {
+  HARD_ASSERT(!path.empty(), "Cannot set field for empty path on ObjectValue");
+
+  google_firestore_v1_Value* nested_value = value_.get();
+  for (const std::string& segment : path.PopLast()) {
+    google_firestore_v1_MapValue_FieldsEntry* entry =
+        FindEntry(*nested_value, segment);
+    if (!entry) {
+      // If the entry is not found, exit early. There is nothing to delete.
+      return;
+    }
+    nested_value = &entry->value;
+  }
+
+  // We can only delete a leaf entry if its parent is a map.
+  if (nested_value->which_value_type ==
+      google_firestore_v1_Value_map_value_tag) {
+    std::set<std::string> deletes{path.last_segment()};
+    ApplyChanges(&nested_value->map_value, /*upserts=*/{}, deletes);
+  }
+}
+
+/**
+ * Returns the map that contains the leaf element of `path`. If the parent
+ * entry does not yet exist, or if it is not a map, a new map will be created.
+ */
+google_firestore_v1_MapValue* MutableObjectValue::ParentMap(
+    const FieldPath& path) {
+  google_firestore_v1_Value* parent = value_.get();
+
+  // Find a or create a parent map entry for `path`.
+  for (const std::string& segment : path) {
+    google_firestore_v1_MapValue_FieldsEntry* entry =
+        FindEntry(*parent, segment);
+
+    if (entry) {
+      if (entry->value.which_value_type !=
+          google_firestore_v1_Value_map_value_tag) {
+        // Since the element is not a map value, free all existing data and
+        // change it to a map type
+        FreeNanopbMessage(FieldsArray<google_firestore_v1_Value>(),
+                          &entry->value);
+        entry->value.which_value_type = google_firestore_v1_Value_map_value_tag;
+      }
+
+      parent = &entry->value;
+    } else {
+      // Create a new map value for the current segment
+      google_firestore_v1_Value new_entry{};
+      new_entry.which_value_type = google_firestore_v1_Value_map_value_tag;
+
+      std::map<std::string, google_firestore_v1_Value> upserts{
+          {segment, new_entry}};
+      ApplyChanges(&(parent->map_value), upserts, /*deletes=*/{});
+
+      parent = &(FindEntry(*parent, segment)->value);
+    }
+  }
+
+  return &parent->map_value;
+}
+
+void MutableObjectValue::ApplyChanges(
```

Comment:
Does this need to be a member function? It doesn't seem to refer to any member variables -- if that's the case, consider making it a free function.

Reply:
It doesn't. I moved it and removed it from the header.